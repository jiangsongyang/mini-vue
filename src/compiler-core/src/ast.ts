import { CREATE_ELEMENT_VNODE } from './runtimeHelpers'

export const enum NodeType {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  TEXT,
  ROOT,
  COMPOUND_EXPRESSION,
}

export function createVNodeCall(context, tag, props, children) {
  context.helper(CREATE_ELEMENT_VNODE)
  return {
    type: NodeType.ELEMENT,
    tag,
    props,
    children,
  }
}