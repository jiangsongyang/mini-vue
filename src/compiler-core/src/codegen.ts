import { isArr, isStr } from '../../shared'
import { NodeType } from './ast'
import {
  CREATE_ELEMENT_VNODE,
  HelperNameMapping,
  TO_DISPLAY_STRING,
} from './runtimeHelpers'

export function codegen(ast) {
  const context = createCodegenContext()
  const { push } = context

  if (ast.helpers.length) {
    genFunctionPreamble(ast, context)
  }

  const funcName = 'render'
  push(`return `)
  const args = ['_ctx', '_cache']
  const signature = args.join(', ')
  push(`function ${funcName}(${signature}) { `)
  push(`return `)
  genNode(ast.codegenNode, context)
  push(` }`)
  return context.code
}

function genFunctionPreamble(ast, context) {
  const VueBinding = 'Vue'
  const { push, newLine } = context
  const aliasHelper = s => `${HelperNameMapping[s]}: _${HelperNameMapping[s]}`
  push(`const { ${ast.helpers.map(aliasHelper).join(', ')} } = ${VueBinding}`)
  newLine()
}

function genNode(node, context) {
  switch (node.type) {
    case NodeType.TEXT:
      genText(node, context)
      break
    case NodeType.INTERPOLATION:
      genInterpolation(node, context)
      break
    case NodeType.SIMPLE_EXPRESSION:
      genExpression(node, context)
      break
    case NodeType.ELEMENT:
      genElement(node, context)
      break
    case NodeType.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context)
  }
}

function genCompoundExpression(node, context) {
  const { children } = node
  const { push } = context
  for (let i = 0; i < children.length; i++) {
    const child = children[i]
    if (isStr(child)) {
      push(child)
    } else {
      genNode(child, context)
    }
  }
}

function genElement(node, context) {
  const { push, helper } = context
  const { tag, props } = node
  push(`${helper(CREATE_ELEMENT_VNODE)}(`)
  const { children } = node
  genNodeList(genNullable([tag, props, children]), context)
  push(')')
}

function genNodeList(nodes, context) {
  const { push } = context
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i]
    if (isStr(node)) {
      push(node)
    } else if (isArr(node)) {
      for (let j = 0; j < node.length; j++) {
        const n = node[j]
        genNode(n, context)
      }
    } else {
      genNode(node, context)
    }
    if (i < nodes.length - 1) {
      push(', ')
    }
  }
}

function genNullable(args) {
  return args.map(arg => arg || 'null')
}

function genExpression(node, context) {
  const { push } = context
  push(`${node.content}`)
}

function genInterpolation(node, context) {
  const { push } = context
  push(`${context.helper(TO_DISPLAY_STRING)}(`)
  genNode(node.content, context)
  push(`)`)
}

function genText(node, context) {
  const { push } = context
  push(`'${node.content}'`)
}

function createCodegenContext() {
  const context = {
    code: '',
    push(source: string) {
      context.code += source
    },
    newLine() {
      context.code += '\n'
    },
    helper(name) {
      return `_${HelperNameMapping[name]}`
    },
  }
  return context
}