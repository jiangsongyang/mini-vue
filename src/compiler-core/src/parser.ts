import { NodeType } from './ast'

const enum TagType {
  START,
  END,
}

export function baseParse(content: string) {
  const context = createContext(content)
  return createRoot(parseChildren(context, []))
}

function createContext(content: string) {
  return {
    source: content,
  }
}

function createRoot(children) {
  return {
    children,
    type: NodeType.ROOT,
  }
}

function parseChildren(context: { source: string }, ancestors): any {
  const nodes: any = []
  while (!isEnd(context, ancestors)) {
    let node
    const s = context.source
    if (s.startsWith('{{')) {
      node = parseInterpolation(context)
    } else if (s.startsWith('<') && /[a-z]/i.test(s[1])) {
      node = parseElement(context, ancestors)
    }
    if (!node) {
      node = parseText(context)
    }
    nodes.push(node)
  }
  return nodes
}

function isEnd(context: { source: string }, ancestors) {
  const s = context.source
  // 2. 遇到结束标签
  if (s.startsWith('</')) {
    for (let i = ancestors.length - 1; i >= 0; i--) {
      const tag = ancestors[i].tag
      if (startsWithEndTagOpen(s, tag)) {
        return true
      }
    }
  }
  // 1. source 有值
  return !s
}

function parseInterpolation(context: { source: string }) {
  const openDelimiter = '{{'
  const closeDelimiter = '}}'
  // 将字符串截取为 message}}
  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  )
  // 让后将字符串前面的 {{ 舍弃掉，我们将其称之为【推进】
  advanceBy(context, openDelimiter.length)
  // 获取到 {{}} 中间值的长度
  const rawContentLength = closeIndex - openDelimiter.length
  // 并将中间这个值获取出来
  const rawContent = parseTextData(context, rawContentLength)
  const content = rawContent.trim()
  // 继续【推进】
  advanceBy(context, rawContentLength + closeDelimiter.length)
  return {
    type: NodeType.INTERPOLATION,
    content: {
      type: NodeType.SIMPLE_EXPRESSION,
      content: content,
    },
  }
}

function advanceBy(context, length: number) {
  context.source = context.source.slice(length)
}

function parseElement(context: { source: string }, ancestors): any {
  const element: any = parseTag(context, TagType.START)
  ancestors.push(element)
  element.children = parseChildren(context, ancestors)
  ancestors.pop()
  if (startsWithEndTagOpen(context.source, element.tag)) {
    parseTag(context, TagType.END)
  } else {
    throw new Error(`unclosed tag ${element.tag}`)
  }
  return element
}

function parseTag(context: { source: string }, type: TagType) {
  // i 忽略大小写, ([a-z]*) 作为一个分组
  const match = /^<\/?([a-z]*)/i.exec(context.source)
  const tag = match![1]
  advanceBy(context, match![0].length + 1)
  if (type === TagType.END) return
  return {
    type: NodeType.ELEMENT,
    tag,
  }
}

function parseText(context: { source: string }): any {
  const s = context.source
  const endTokens = ['<', '{{']
  let endIndex = s.length
  for (let i = 0; i < endTokens.length; i++) {
    const index = s.indexOf(endTokens[i])
    if (index !== -1 && endIndex > index) {
      endIndex = index
    }
  }
  const content = parseTextData(context, endIndex)
  advanceBy(context, content.length)
  return {
    type: NodeType.TEXT,
    content,
  }
}

function parseTextData(context: { source: string }, length) {
  return context.source.slice(0, length)
}

function startsWithEndTagOpen(source, tag) {
  const endTokenLength = '</'.length
  return source.slice(endTokenLength, tag.length + endTokenLength) === tag
}