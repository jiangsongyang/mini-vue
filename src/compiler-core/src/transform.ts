import { NodeType } from './ast'
import { HelperNameMapping, TO_DISPLAY_STRING } from './runtimeHelpers'

export function transform(root, options = {}) {
  const context = createTransformContext(root, options)
  traverseNode(root, context)
  createRootCodegen(root)
  root.helpers = [...context.helpers.keys()]
}

function createRootCodegen(root) {
  const child = root.children[0]
  if (child.type === NodeType.ELEMENT) {
    root.codegenNode = child.codegenNode
  } else {
    root.codegenNode = root.children[0]
  }
}

function createTransformContext(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || {},
    helpers: new Map(),
    helper(name: string) {
      context.helpers.set(name, 1)
    },
  }
  return context
}

function traverseNode(node, context) {
  const { nodeTransforms } = context
  const exitFns: any[] = []
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i]
    const exitFn = transform(node, context)
    if (exitFn) exitFns.push(exitFn)
  }
  switch (node.type) {
    case NodeType.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING)
      break
    case NodeType.ROOT:
    case NodeType.ELEMENT:
      traverseChildren(node, context)
      break
    default:
      break
  }
  let i = exitFns.length
  while (i--) {
    exitFns[i]()
  }
}

function traverseChildren(node, context) {
  const children = node.children
  for (let i = 0; i < children.length; i++) {
    traverseNode(children[i], context)
  }
}