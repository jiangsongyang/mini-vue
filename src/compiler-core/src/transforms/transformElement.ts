import { createVNodeCall, NodeType } from '../ast'

export function transformElement(node, context) {
  if (node.type === NodeType.ELEMENT) {
    return () => {
      // 中间处理层，处理 props 和 tag
      const vnodeTag = `'${node.tag}'`
      const vnodeProps = node.props

      const { children } = node
      const vnodeChildren = children

      node.codegenNode = createVNodeCall(
        context,
        vnodeTag,
        vnodeProps,
        vnodeChildren
      )
    }
  }
}