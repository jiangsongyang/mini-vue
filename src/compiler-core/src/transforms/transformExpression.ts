import { NodeType } from '../ast'

export function transformExpression(node) {
  if (node.type === NodeType.INTERPOLATION) {
    node.content = processExpression(node.content)
  }
}

function processExpression(node) {
  node.content = `_ctx.${node.content}`
  return node
}