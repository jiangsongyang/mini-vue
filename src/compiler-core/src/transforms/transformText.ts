import { NodeType } from '../ast'
import { isText } from '../utils'

export function transformText(node) {
  return () => {
    const { children } = node
    if (children && children.length) {
      let currentContainer
      for (let i = 0; i < children.length; i++) {
        const child = children[i]
        if (isText(child)) {
          for (let j = i + 1; j < children.length; j++) {
            const next = children[j]
            if (isText(next)) {
              // 相邻的是 text 或者 interpolation，那么就变成联合类型
              if (!currentContainer) {
                currentContainer = children[i] = {
                  type: NodeType.COMPOUND_EXPRESSION,
                  children: [child],
                }
              }
              // 在每个相邻的下一个之前加上一个 +
              currentContainer.children.push(' + ')
              currentContainer.children.push(next)
              // 遇到就删除
              children.splice(j, 1)
              // 修正索引，因为我们下一个循环就又 + 1 了。此时索引就不对了
              j -= 1
            } else {
              // 如果下一个不是 text 的了，那么就重置，并跳出循环
              currentContainer = undefined
              break
            }
          }
        }
      }
    }
  }
}