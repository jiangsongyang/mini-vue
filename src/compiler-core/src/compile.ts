import { codegen } from './codegen'
import { baseParse } from './parser'
import { transform } from './transform'
import { transformElement } from './transforms/transformElement'
import { transformExpression } from './transforms/transformExpression'
import { transformText } from './transforms/transformText'

export function baseCompile(template: string) {
  const ast = baseParse(template)
  transform(ast, {
    nodeTransforms: [transformExpression, transformElement, transformText],
  })
  const code = codegen(ast)
  return {
    code,
  }
}