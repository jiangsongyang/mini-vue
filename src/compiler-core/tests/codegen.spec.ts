import { codegen } from '../src/codegen'
import { baseParse } from '../src/parser'
import { transform } from '../src/transform'
import { transformElement } from '../src/transforms/transformElement'
import { transformExpression } from '../src/transforms/transformExpression'
import { transformText } from '../src/transforms/transformText'

describe('codegen', () => {
  test('text', () => {
    const template = 'hi'
    const ast = baseParse(template)
    transform(ast)
    const code = codegen(ast)
    expect(code).toMatchSnapshot()
  })
  test('interpolation', () => {
    const template = '{{message}}'
    const ast = baseParse(template)
    transform(ast, {
      nodeTransforms: [transformExpression],
    })
    const code = codegen(ast)
    expect(code).toMatchSnapshot()
  })
  test('simple element', () => {
    const template = '<div></div>'
    const ast = baseParse(template)
    transform(ast, {
      nodeTransforms: [transformElement],
    })
    const code = codegen(ast)
    expect(code).toMatchSnapshot()
  })
  test('union 3 type', () => {
    const template = '<div>hi,{{message}}</div>'
    const ast = baseParse(template)
    transform(ast, {
      nodeTransforms: [transformExpression, transformElement, transformText],
    })
    const code = codegen(ast)
    expect(code).toMatchSnapshot()
  })
})