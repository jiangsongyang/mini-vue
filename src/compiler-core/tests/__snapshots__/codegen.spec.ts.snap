// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen interpolation 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache) { return _toDisplayString(_ctx.message) }"
`;

exports[`codegen simple element 1`] = `
"const { createElementVNode: _createElementVNode } = Vue
return function render(_ctx, _cache) { return _createElementVNode('div', null, ) }"
`;

exports[`codegen text 1`] = `"return function render(_ctx, _cache) { return 'hi' }"`;

exports[`codegen union 3 type 1`] = `
"const { toDisplayString: _toDisplayString, createElementVNode: _createElementVNode } = Vue
return function render(_ctx, _cache) { return _createElementVNode('div', null, 'hi,' + _toDisplayString(_ctx.message)) }"
`;
